### PR TITLE
Asoc: SOF: loader: check for unsigned wrap in subtraction

### LINKED ISSUE #793: Upstream: v5: loader code hardening
Feedback from Takashi Iwai

````
> +		if (remaining < block->size) {
> +			dev_err(sdev->dev, "error: not enough data remaining\n");
> +			return -EINVAL;
> +		}

remaining is unsigned, so a negative check doesn't work here.
Hence you need the explicit underflow check.
````

### PR DESCRIPTION
**Pull request by juimonen**
fix a check with possible unsigned wrap based on feedback
from Takashi Iwai.

Signed-off-by: Jaska Uimonen <jaska.uimonen@intel.com>

fixes #793

### CODE REVIEW

File: sound/soc/sof/loader.c [hunk 1]
```diff
@@ -203,6 +209,12 @@ static int load_modules(struct snd_sof_dev *sdev, const struct firmware *fw)
 	module = (struct snd_sof_mod_hdr *)((u8 *)(fw->data) + sizeof(*header));
 	remaining = fw->size - sizeof(*header);
 	for (count = 0; count < header->num_modules; count++) {
+		/* check for wrap */
```
Comment (plbossart) [inline sound/soc/sof/loader.c:219]: remaining can already wrap two lines above?

Comment (plbossart) [reply sound/soc/sof/loader.c:219]: and it can also wrap below (line 231)

File: sound/soc/sof/loader.c [hunk 2]
```diff
@@ -106,6 +106,12 @@ int snd_sof_parse_module_memcpy(struct snd_sof_dev *sdev,
 	/* module->size doesn't include header size */
 	remaining = module->size;
 	for (count = 0; count < module->num_blocks; count++) {
+		/* check for wrap */
+		if (remaining < sizeof(*block)) {
+			dev_err(sdev->dev, "error: not enough data remaining\n");
+			return -EINVAL;
+		}
+
```
Comment (plbossart) [inline sound/soc/sof/loader.c:114]: i'd move the test line 117 to line 156. The code as is looks like a mistaken copy-paste, when it's really doing the check earlier than needed (and other errors could be trapped first before decrementing line 156)

File: sound/soc/sof/loader.c (hunk 1 above)
Comment (juimonen) [reply sound/soc/sof/loader.c:219]: ok

File: sound/soc/sof/loader.c (hunk 2 above)
Comment (juimonen) [reply sound/soc/sof/loader.c:114]: ok

**Comment by juimonen** [comment]
@plbossart updated. Hope I didn't change the logic of the loader... check if you spot something, I 'll fix tomorrow morning my time.

File: sound/soc/sof/loader.c [hunk 3]
```diff
@@ -202,19 +209,34 @@ static int load_modules(struct snd_sof_dev *sdev, const struct firmware *fw)
 	/* parse each module */
 	module = (struct snd_sof_mod_hdr *)((u8 *)(fw->data) + sizeof(*header));
 	remaining = fw->size - sizeof(*header);
+	/* check for wrap */
+	if (remaining > fw->size) {
```
Comment (plbossart) [inline sound/soc/sof/loader.c:213]: this looks wrong. We should first check that fw->size >= sizeof(*header).
I didn't spot any other issues.

Comment (juimonen) [reply sound/soc/sof/loader.c:213]: @plbossart I tried to do "compliant post condition test for unsigned subtraction", like here:
https://wiki.sei.cmu.edu/confluence/display/c/INT30-C.+Ensure+that+unsigned+integer+operations+do+not+wrap
looks little bit strange, but should work...

Comment (kv2019i) [reply sound/soc/sof/loader.c:213]: Ack, I think it's indeed correct -- the comment is good though.

Comment (plbossart) [reply sound/soc/sof/loader.c:213]: Ah, thanks for teaching me something :-)

For the record this is what @juimonen used:

````c
void func(unsigned int ui_a, unsigned int ui_b) {
  unsigned int udiff = ui_a - ui_b;
  if (udiff > ui_a) {
    /* Handle error */
  }
  /* ... */
}
````

**Pull request merged**